src/enumerable.cr

```diff
@@ -1205,6 +1205,50 @@ module Enumerable(T)
     reduce(initial) { |memo, e| memo * (yield e) }
   end
 
+  # Returns an array of tuples containing the cartesian product
+  # between the collection and the enumerable. Or in other words, it
+  # will return an Array with a Tuple for each possible combination of
+  # two elements from the collection and the passed enumerable.
+  #
+  # ```
+  # arr = [1, 2, 3]
+  # set = Set(Char).new(['a', 'b'])
+  # arr.product(set)
+  # => [{1, 'a'}, {1, 'b'}, {2, 'a'}, {2, 'b'}, {3, 'a'}, {3, 'b'}])
+  # ```
+  #
+  # An empty array will be returned if either collection is empty.
+  def product(ary : Enumerable(U)) forall U
+    result = if ary.responds_to?(:size)
```

Comment:
Ouch, that would be a bad side effect without any obvious solution. Should I close or leave the pull request?